### PR TITLE
HEEDLS-653 - Add Delegate courses precursor page

### PR DESCRIPTION
**Pull request by GeorgeBlackwell**
### JIRA link
[_HEEDLS-653_](https://softwiretech.atlassian.net/browse/HEEDLS-653)

### Description
There was already a [MR](https://github.com/TechnologyEnhancedLearning/DLSV2/pull/948) but I messed up the branch/commit history due to habits from a previous project. This has all the changes from the previous branch and the required mark-ups.

A new precursor page has been added to _Course delegates_. When a user attempts to access this page, they'll be taken to _Delegate courses_ before moving onto the existing page.

### Screenshots
image
image
image
image
image

-----
### Developer checks
(Leave tasks unticked if they haven't been appropriate for your ticket.)

I have:
- [X] Run the formatter and made sure there are no IDE errors.
- [X] Written tests for the changes (accessibility tests, unit tests for controller, data services, services, view models, etc)
- [X] Manually tested my work with and without JavaScript. Full manual testing guidelines can be found here: https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/6703648740/Testing
- [ ] Updated/added documentation in Swiki and/or Readme. Links (if any) are below:
  - [doc_1_here](link_1_here)
- [X] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken.
- [X] Scanned over my own MR to ensure everything is as expected.

### CODE REVIEW

File: DigitalLearningSolutions.Web.Tests/Controllers/TrackingSystem/Delegates/DelegateCoursesControllerTests.cs
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Web.Tests/Controllers/TrackingSystem/Delegates/DelegateCoursesControllerTests.cs:76]: Just to clarify, on the old PR, Steve had a question about setting up the cookies inside the tests to make the individual tests read slightly easier. Was this decided against?

File: DigitalLearningSolutions.Web.Tests/Controllers/TrackingSystem/Delegates/DelegateCoursesControllerTests.cs
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Web.Tests/Controllers/TrackingSystem/Delegates/DelegateCoursesControllerTests.cs:54]: These should probably be changed to match the new CourseService method that is being called.

File: DigitalLearningSolutions.Web/Views/TrackingSystem/Centre/TopCourses/Index.cshtml
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Web/Views/TrackingSystem/Centre/TopCourses/Index.cshtml:11]: Lower on this page, the `View more` button still needs hooking up to the new page instead of the old Course Delegates page.

File: DigitalLearningSolutions.Web.Tests/ViewModels/Common/SearchablePageViewModelTests.cs
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Web.Tests/ViewModels/Common/SearchablePageViewModelTests.cs:13]: Nice.

**Review by DanBloxham-sw — CHANGES_REQUESTED** [review]
I've left a few follow-up comments

Asked a question to @SteveJacksonSoft on the old MR about the purpose of the page, might make everything here redundant if my thinking is correct, so don't know if you want to wait for an answer to that before changing anything.

File: DigitalLearningSolutions.Web/Views/TrackingSystem/Delegates/DelegateCourses/Index.cshtml
Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Web/Views/TrackingSystem/Delegates/DelegateCourses/Index.cshtml:5]: Thanks for pointing that out - nah, I don't think it's worth it.

File: DigitalLearningSolutions.Web.Tests/Controllers/TrackingSystem/Delegates/DelegateCoursesControllerTests.cs
Comment (GeorgeBlackwell) [reply DigitalLearningSolutions.Web.Tests/Controllers/TrackingSystem/Delegates/DelegateCoursesControllerTests.cs:76]: It proved to be a bit of pain to implement and considering its required in a good number of the tests it made sense to leave it where it is.

File: DigitalLearningSolutions.Data/Services/CourseService.cs
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Data/Services/CourseService.cs:337]: Minor, but might as well make a constructor to assign these.

**Review by DanBloxham-sw — CHANGES_REQUESTED** [review]
One very minor comment, but otherwise great.

File: DigitalLearningSolutions.Data/Services/CourseService.cs
Comment (SteveJacksonSoft) [reply DigitalLearningSolutions.Data/Services/CourseService.cs:337]: I agree, we use nullability a fair bit on this project, so it's good if we can ensure that properties are always initialized if they're not marked as nullable.

**Review by DanBloxham-sw — APPROVED** [review]
Looks great.

**Review by SteveJacksonSoft — COMMENTED** [review]
Nice. Looks like some tests are failing though?

File: DigitalLearningSolutions.Web.Tests/ViewModels/Common/SearchablePageViewModelTests.cs
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Web.Tests/ViewModels/Common/SearchablePageViewModelTests.cs]: Does adding `.OrderBy(stats => stats.ApplicationName)` in here fix things?

Comment (GeorgeBlackwell) [reply DigitalLearningSolutions.Web.Tests/ViewModels/Common/SearchablePageViewModelTests.cs]: Sorry, just seen this after the merge. There was a lot inconsistency with the test passing locally and then failing in Jenkins, the list was seemly being generated differently. It seems to be the same issue Alex is having in [790](https://github.com/TechnologyEnhancedLearning/DLSV2/pull/974/files#r814782462)

**Pull request merged**
